gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/DistributedIdentifyUtil.java

```diff
@@ -0,0 +1,309 @@
+    return identifiedResult;
+  }
+
+  protected Map<String, ContentBloomFilter> computeLiveContents(
+      Map<String, String> config,
+      Map<String, Instant> cutOffTimestampPerRef,
+      Instant defaultCutOffTimestamp,
+      Reference reference) {
+    try (NessieApiV1 api = getApi(config)) {
+      Instant cutOffTimestamp =
+          cutOffTimestampPerRef.getOrDefault(reference.getName(), defaultCutOffTimestamp);
+      return walkLiveReference(
+          api,
+          reference,
+          commitMeta ->
+              // If the commit time is newer than (think: greater than or equal to) cutoff-time,
+              // then
+              // commit is live.
+              Objects.requireNonNull(commitMeta.getCommitTime()).compareTo(cutOffTimestamp) >= 0);
+    }
+  }
+
+  protected @NotNull IdentifiedResult computeExpiredContents(
+      Map<String, String> config,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap,
+      Reference reference) {
+    try (NessieApiV1 api = getApi(config)) {
+      return walkAllReference(api, reference, liveContentsBloomFilterMap);
+    }
+  }
+
+  private int getTaskCount(List<Reference> references) {
+    int taskCount =
+        Integer.parseInt(
+            config.getOrDefault("nessie.gc.spark.task.count", String.valueOf(references.size())));
+    if (taskCount <= 0) {
+      LOGGER.warn("Invalid nessie.gc.spark.task.count configuration: {} using default", taskCount);
+      taskCount = references.size();
+    }
+    return taskCount;
+  }
+
+  private Map<String, ContentBloomFilter> mergeLiveContentResults(
+      List<Map<String, ContentBloomFilter>> bloomFilterMaps) {
+    Map<String, ContentBloomFilter> output = new HashMap<>();
+    bloomFilterMaps.forEach(
+        map ->
+            map.forEach(
+                (k, v) -> {
+                  if (output.containsKey(k)) {
+                    output.get(k).merge(v);
+                  } else {
+                    output.put(k, v);
+                  }
+                }));
+    return output;
+  }
+
+  private Map<String, ContentBloomFilter> walkLiveReference(
+      NessieApiV1 api, Reference reference, Predicate<CommitMeta> liveCommitPredicate) {
+    Map<String, ContentBloomFilter> bloomFilterMap = new HashMap<>();
+    Set<ContentKey> deletedKeys = new HashSet<>();
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+      Spliterator<LogResponse.LogEntry> src = commits.spliterator();
+      // Use a Spliterator to limit the processed commits to the "live" commits - i.e. stop at the
+      // first "non-live" commit.
+      new Spliterators.AbstractSpliterator<LogResponse.LogEntry>(src.estimateSize(), 0) {
+        private boolean more = true;
+        private boolean refHead = true;
+
+        @Override
+        public boolean tryAdvance(Consumer<? super LogResponse.LogEntry> action) {
```

Comment:
This Spliterator should really be unit tested and live in a separate utility class. It's too easy to miss cases here, that are afterwards hard to detect / debug.

Reply:
done. moved to a Util class and added a UT.